### ISSUE TITLE
user keys still active after user logout

### ISSUE THREAD

**Issue by dkg** [systemd/systemd#33308]
### systemd version the issue has been seen with

256-1

### Used distribution

debian unstable

### Linux kernel version used

6.8.12-amd64

### CPU architectures issue was seen on

x86_64

### Component

systemd-homed, systemd-logind

### Expected behaviour you didn't see

Following up from #33138 and #32596, there are still some keys lingering in the kernel after the user has fully logged out.

The user `moogly` has a home directory made with `fscrypt`, does not have lingering configured, and logs in from the virtual terminal, and logs back out again.

I would have expected `/proc/key-user` and `/proc/keys` to return to the same state that it started in after logout.

### Unexpected behaviour you saw

Before user `moogly` logs in, i see this:

```
0 root@unstable-amd64:~# chpst -u :60240:60240 cat /proc/keys && cat /proc/key-users
25f1dfe0 I--Q---    14 perm 3f030000     0     0 keyring   _ses: 1
38b24588 I--Q---     4 perm 1f3f0000     0 65534 keyring   _uid.0: empty
    0:    47 46/46 30/1000000 577/25000000
  101:     1 1/1 1/200 9/20000
  102:     2 2/2 2/200 18/20000
0 root@unstable-amd64:~# 
```

10 minutes after `moogly` logs out, i see this:

```
0 root@unstable-amd64:~# chpst -u :60240:60240 cat /proc/keys && cat /proc/key-users 
0ada60c6 I--Q---     1 perm 1f3f0000 60240 65534 keyring   _uid_ses.60240: 1
0f2c585b I--Q---     2 perm 1f3f0000 60240 65534 keyring   _uid.60240: empty
25f1dfe0 I--Q---    14 perm 3f030000     0     0 keyring   _ses: 1
38b24588 I--Q---     4 perm 1f3f0000     0 65534 keyring   _uid.0: empty
    0:    47 46/46 30/1000000 585/25000000
  101:     1 1/1 1/200 9/20000
  102:     2 2/2 2/200 18/20000
60240:     2 2/2 2/200 30/20000
0 root@unstable-amd64:~#     
```

### Steps to reproduce the problem

same approach as in #33138 and #32596

**Comment by bluca** [comment]
I don't think this is to do with systemd?

**Comment by dkg** [comment]
I depend on systemd to load the user's keys into the kernel when their session starts, and to ensures that the user's session is fully closed (including key removal) when they leave.  If it's a bug purely in the kernel, i ought to be able to replicate it with something unrelated to systemd, right?  Can you suggest a path for such a replication?

**Comment by poettering** [comment]
homed really only loads the fscrypt key

**Comment by poettering** [comment]
I mean, as mentioned we can look into explicitly clearing the keyring on logout, we do so for IPC objects after all too, but the kernel should theoreticall do this on its own

**Comment by poettering** [comment]
So these two entries are "left over" from your session:

```
0ada60c6 I--Q---     1 perm 1f3f0000 60240 65534 keyring   _uid_ses.60240: 1
0f2c585b I--Q---     2 perm 1f3f0000 60240 65534 keyring   _uid.60240: empty
```

These two are not keys, but keyrings. i.e. they do not carry any key material themselves, but just contain references to other keys. (in other words: if keys are "files", then keyrings are "directories").

The first one is the `user-session` keyring of your user, see user-session-keyring(7) for details.

There it says:

> The  keyring remains pinned in existence so long as there are processes running with that real UID or files opened by those processes remain open.  (The keyring can also be pinned indefinitely by linking it into another keyring.)

The other one is the `user` keyring of your user, see user-keyring(7) for details. It says:

> It comes into existence upon the first attempt to access either the user keyring, the user-session-keyring(7), or the  session-keyring(7). The  keyring  remains  pinned  in  existence  so  long as there are processes running with that real UID or files opened by those processes remain open.  (The keyring can also be pinned indefinitely by linking it into another       keyring.)

So, are you sure you have no process running anymore under that UID? And no fds open from that user either? (that was passed elsewhere?)

**Comment by dkg** [comment]
> So, are you sure you have no process running anymore under that UID? And no fds open from that user either? (that was passed elsewhere?)

Yes, i'm certain that there are no processes bound to that UID.  As for fds open from that user, i can't see how it would happen, but i don't know how to exhaustively search.

The only thing the user did was:

- log in on a virtual console to a text-mode session
- `cat test.txt`
- log out

You should be able to replicate this with a minimal virtual machine.
